cmd/csi-attacher/main.go

```diff
@@ -170,30 +188,8 @@ func main() {
 		}
 		// Name of config map with leader election lock
 		lockName := "external-attacher-leader-" + attacher
-		waitForLeader(clientset, *leaderElectionNamespace, *leaderElectionIdentity, lockName)
+		runAsLeader(clientset, *leaderElectionNamespace, *leaderElectionIdentity, lockName, run)
 	}
-
-	ctrl := controller.NewCSIAttachController(
-		clientset,
-		attacher,
-		handler,
-		factory.Storage().V1beta1().VolumeAttachments(),
-		factory.Core().V1().PersistentVolumes(),
-	)
-
-	// run...
-	stopCh := make(chan struct{})
-	factory.Start(stopCh)
-	if csiFactory != nil {
-		csiFactory.Start(stopCh)
-	}
-	go ctrl.Run(threads, stopCh)
-
-	// ...until SIGINT
```

Comment:
This piece got lost in this PR. IMO it makes some sense to wait until processing of an event is finished and not interrupt it in the middle.

Reply:
I am a bit confused if this works. The program will exit immediately after `close(stopCh)` finished, as closing of a channel happens before a receive that returns a zero value because the channel is closed, more details can be found [here](https://golang.org/ref/mem#tmp_7). That means all goroutines will exit immediately too.

At this moment the `CSIAttachController` may be processing an event and will still be interrupted in the middle.

If we want to wait until processing of an event is finished, maybe we need to add another channel which will be closed when these two [sync funcs](https://github.com/kubernetes-csi/external-attacher/blob/068527da76a9cfc18c4bae08d1d453c6f89fabbd/pkg/controller/controller.go#L118) finished and the `main` function should wait for this channel before exiting. But this may make the code too complicated